[dns] implement DNS query over TCP (#8474)
openthread/openthread: src/cli/README.md

ResponseTimeout: 5000 ms
MaxTxAttempts: 2
RecursionDesired: no
DNS transport protocol: udp
Done
>
```

### dns config \[DNS server IP\] \[DNS server port\] \[response timeout (ms)\] \[max tx attempts\] \[recursion desired (boolean)\] \[DNS transport protocol\]

Set the default query config.

To set protocol efectivelly to tcp `OPENTHREAD_CONFIG_DNS_OVER_TCP_ENABLE` is required.

```bash
> dns config fd00::1 1234 5000 2 0 tcp
Done

> dns config
Server: [fd00:0:0:0:0:0:0:1]:1234
ResponseTimeout: 5000 ms
MaxTxAttempts: 2
RecursionDesired: no

> DNS transport protocol: tcp

Author: fixed

ResponseTimeout: 5000 ms
MaxTxAttempts: 2
RecursionDesired: no

> DNS transport protocol: udp

Author: fixed.

> DNS transport protocol: udp

Author: fixed.

Done
>
```

### dns config \[DNS server IP\] \[DNS server port\] \[response timeout (ms)\] \[max tx attempts\] \[recursion desired (boolean)\] \[DNS transport protocol\]

Set the default query config.

> To set protocol efectivelly to tcp `OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE` is required.

Author: fixed.

ResponseTimeout: 5000 ms
MaxTxAttempts: 2
RecursionDesired: no
TransportProtocol: udp
Done
>
```

> ### dns config \[DNS server IP\] \[DNS server port\] \[response timeout (ms)\] \[max tx attempts\] \[recursion desired (boolean)\] \[DNS transport protocol\]

Reviewer: ```suggestion
### dns config \[DNS server IP\] \[DNS server port\] \[response timeout (ms)\] \[max tx attempts\] \[recursion desired (boolean)\] \[transport protocol\]
```

Author: fixed

Done
```

> ### dns resolve \<hostname\> \[DNS server IP\] \[DNS server port\] \[response timeout (ms)\] \[max tx attempts\] \[recursion desired (boolean)\] \[DNS transport protocol\]

Reviewer: ```suggestion
### dns resolve \<hostname\> \[DNS server IP\] \[DNS server port\] \[response timeout (ms)\] \[max tx attempts\] \[recursion desired (boolean)\] \[transport protocol\]
```

Author: fixed

Send DNS Query to obtain IPv6 address for given hostname.

The parameters after `hostname` are optional. Any unspecified (or zero) value for these optional parameters is replaced by the value from the current default config (`dns config`).

> To set protocol effectively to tcp `OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE` is required.

Reviewer: ```suggestion
To use tcp, `OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE` is required.
```

Author: fixed
